# 01. Import Libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [2]:
path = r"C:\Users\edriesberg\OneDrive - Cretex Companies, Inc\Personal\Data Analytics\Achievement 4\Instacart Basket Analysis"

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_merged.pkl'))

In [4]:
df = ords_prods_merge[:1000000]

In [5]:
def price_label(row):
    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High-range'
    else: return 'Not enough data'

In [6]:
df['price_range'] = df.apply(price_label, axis = 1)

<ipython-input-6-ea42c94d65ec>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [7]:
df['price_range'].value_counts()

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [8]:
df['prices'].max()

14.8

In [9]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [10]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [11]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [12]:
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

In [13]:
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [14]:
result = []

for value in ords_prods_merge['orders_day_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy')
    else:
        result.append('Regularly busy')

In [15]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [16]:
ords_prods_merge['busiest day'] = result

In [17]:
ords_prods_merge['busiest day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest day, dtype: int64

## Question 2.
Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [18]:
result = []

for value in ords_prods_merge['orders_day_of_week']:
    if value == 0 or value == 1:
        result.append('Most busy')
    elif value == 4 or value == 3:
        result.append('Least busy')
    else:
        result.append('Regularly busy')

In [19]:
ords_prods_merge['busiest days'] = result

## Question 3.
Check the values of this new column for accuracy. Note any observations in markdown format.

In [20]:
ords_prods_merge['busiest days'].value_counts(dropna = False)

Regularly busy    12916111
Most busy         11864412
Least busy         7624336
Name: busiest days, dtype: int64

Most busy = 6204182 + 5660230 = 11864412

Least busy = 3783802 + 3840534 = 7624336

Remainder = 12916111

Total observations = 11864412 + 7624336 + 12916111 = 32404859

## Question 4.
When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [21]:
ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [22]:
hours = []

busy_time = [10,11,14,15,13,12,16,9]
slow_time = [3,4,2,5,1,0,6,23]

for x in ords_prods_merge['order_hour_of_day']:
    if x in busy_time:
        hours.append('Most orders')
    elif x in slow_time:
        hours.append('Fewest orders')
    else:
        hours.append('Average orders')
    

In [23]:
ords_prods_merge['busiest_period_of_day'] = hours

## Question 5.
Print the frequency for this new column.

In [24]:
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: busiest_period_of_day, dtype: int64

In [25]:
ords_prods_merge.to_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_new_vars.pkl'))